core/src/main/scala/kafka/controller/ControllerRequestMerger.scala

```diff
@@ -0,0 +1,245 @@
+import org.apache.kafka.common.message.LiCombinedControlRequestData
+import org.apache.kafka.common.message.LiCombinedControlRequestData._
+import org.apache.kafka.common.requests._
+import org.apache.kafka.common.utils.LiCombinedControlTransformer
+import org.apache.kafka.common.{Node, TopicPartition}
+
+import java.util
+import scala.collection.mutable
+
+case class RequestControllerState(controllerId: Int, controllerEpoch: Int)
+
+class ControllerRequestMerger extends Logging {
+  val leaderAndIsrPartitionStates: mutable.Map[TopicPartition, util.LinkedList[LeaderAndIsrPartitionState]] = mutable.HashMap.empty
+  var leaderAndIsrLiveLeaders: util.Collection[Node] = new util.ArrayList[Node]()
+
+  val updateMetadataPartitionStates: mutable.Map[TopicPartition, UpdateMetadataPartitionState] = mutable.HashMap.empty
+  var updateMetadataLiveBrokers: util.List[UpdateMetadataBroker] = new util.ArrayList[UpdateMetadataBroker]()
+
+  val stopReplicaPartitionStates: mutable.Map[TopicPartition, util.LinkedList[StopReplicaPartitionState]] = mutable.HashMap.empty
+
+  // If a controller resigns and becomes the active controller again, a new
+  // ControllerRequestMerger object will be created for each RequestSendThread.
+  // Thus the controllerState, once set, should not change for the lifetime of this object.
+  var controllerState : RequestControllerState = null
+
+  // Here we store one callback for the LeaderAndIsr response and one for the StopReplica response,
+  // given all the requests of a particular type sent to the same broker always have the same callback.
+  var leaderAndIsrCallback: AbstractResponse => Unit = null
+  var stopReplicaCallback: AbstractResponse => Unit = null
+
+  def addRequest(request: AbstractControlRequest.Builder[_ <: AbstractControlRequest],
+    callback: AbstractResponse => Unit = null): Unit = {
+    val newControllerState = new RequestControllerState(request.controllerId(), request.controllerEpoch())
+    if (controllerState != null) {
+      if (!controllerState.equals(newControllerState)) {
+        throw new IllegalStateException("The controller state in the ControllerRequestMerger should not change")
+      }
+    } else {
+      controllerState = newControllerState
+    }
+
+    request match {
+      case leaderAndIsrRequest : LeaderAndIsrRequest.Builder => addLeaderAndIsrRequest(leaderAndIsrRequest, callback)
+      case updateMetadataRequest : UpdateMetadataRequest.Builder => addUpdateMetadataRequest(updateMetadataRequest)
+      case stopReplicaRequest: StopReplicaRequest.Builder => addStopReplicaRequest(stopReplicaRequest, callback)
+    }
+  }
+
+  def isLeaderAndIsrReplaceable(newState: LeaderAndIsrPartitionState, currentState: LeaderAndIsrPartitionState): Boolean = {
+    newState.brokerEpoch() > currentState.brokerEpoch() || newState.leaderEpoch() > currentState.leaderEpoch()
+  }
+
+  def mergeLeaderAndIsrPartitionState(incomingState: LeaderAndIsrPartitionState,
```

Comment:
it is a replacement of the previous state or inserts a new state if the previous state cannot be replaced. "merge" is a little bit misleading here.